src/compiler/types.rs

```diff
@@ -41,19 +113,84 @@ impl NodeInterface for Node {
             Node::BaseNode(base_node) => base_node.kind(),
             Node::Expression(expression) => expression.kind(),
             Node::Statement(statement) => statement.kind(),
+            Node::SourceFile(source_file) => source_file.kind(),
+        }
+    }
+
+    fn parent(&self) -> Rc<Node> {
+        match self {
+            Node::BaseNode(base_node) => base_node.parent(),
+            Node::Expression(expression) => expression.parent(),
+            Node::Statement(statement) => statement.parent(),
+            Node::SourceFile(source_file) => source_file.parent(),
+        }
+    }
+
+    fn set_parent(&self, parent: Rc<Node>) {
+        match self {
+            Node::BaseNode(base_node) => base_node.set_parent(parent),
+            Node::Expression(expression) => expression.set_parent(parent),
+            Node::Statement(statement) => statement.set_parent(parent),
+            Node::SourceFile(source_file) => source_file.set_parent(parent),
         }
     }
 }
 
 #[derive(Debug)]
 pub struct BaseNode {
     pub kind: SyntaxKind,
+    pub parent: RwLock<Option<Weak<Node>>>,
+    pub pos: AtomicUsize,
```

Comment:
I used `AtomicUsize`'s here so that `set_pos()`/`set_end()` could take `&self` rather than `&mut self`, otherwise I would've been in a situation where since the `Node` has already been wrapped in an `Rc` (when we want to set its `pos`/`end`) it doesn't want to give us mutable access to its contents